privacyidea/lib/tokens/vascotoken.py

```diff
@@ -0,0 +1,141 @@
+from privacyidea.lib.tokenclass import TokenClass
+from privacyidea.lib.tokens.vasco import vasco_otp_check
+from privacyidea.lib import _
+
+optional = True
+required = False
+
+log = logging.getLogger(__name__)
+
+
+class VascoTokenClass(TokenClass):
+    """
+    Token class for VASCO Digipass tokens. Relies on vendor-specific
+    shared library, whose location needs to be set in the VASCO_LIBRARY
+    config option.
+    """
+
+    def __init__(self, db_token):
+        """
+        constructor - create a token class object with its db token binding
+
+        :param aToken: the db bound token
+        """
+        TokenClass.__init__(self, db_token)
+        self.set_type(u"vasco")
+        self.hKeyRequired = True
+
+    @staticmethod
+    def get_class_type():
+        """
+        return the class type identifier
+        """
+        return "vasco"
+
+    @staticmethod
+    def get_class_prefix():
+        """
+        return the token type prefix
+        """
+        return "VASC"
+
+    @staticmethod
+    @log_with(log)
+    def get_class_info(key=None, ret='all'):
+        """
+        :param key: subsection identifier
+        :type key: string
+        :param ret: default return value, if nothing is found
+        :type ret: user defined
+        :return: subsection if key exists or user defined
+        :rtype: dict or string
+        """
+        res = {'type': 'vasco',
+               'title': 'VASCO Token',
+               'description': _('VASCO Token: Authentication using VASCO tokens'),
+               'user': ["enroll"],
+               'ui_enroll': ["admin", "user"],
+               'policy': {},
+               }
+
+        if key:
+            ret = res.get(key, {})
+        else:
+            if ret == 'all':
+                ret = res
+        return ret
+
+    @log_with(log)
+    def update(self, param, reset_failcount=True):
+        """
+        update - process initialization parameters
+
+        :param param: dict of initialization parameters
+        :type param: dict
+
+        :return: nothing
+        """
+        if is_true(getParam(param, 'genkey', optional)):
+            raise ParameterError("Generating OTP keys is not supported")
+
+        upd_param = param.copy()
+
+        # If the OTP key is given, it is given as a 496-character hex string which
+        # encodes a 248-byte blob. As we want to set a 248-byte OTPKey (= Blob),
+        # we unhexlify the OTP key
+        if 'otpkey' in param:
+            upd_param['otpkey'] = binascii.unhexlify(upd_param['otpkey'])
+
+        TokenClass.update(self, upd_param, reset_failcount)
+
+    @check_token_locked
+    def check_otp(self, otpval, counter=None, window=None, options=None):
+        secret = self.token.get_otpkey().getKey()
+        result, new_secret = vasco_otp_check(secret, otpval)
+        self.token.set_otpkey(new_secret)
+        self.save()
+
+        if result == 0:
+            # Successful authentication
+            return 0
```

Comment:
@cornelinux Is a return value of ``0`` ok here? (I think HOTP tokens return the correct OTP value?)

Reply:
our privacyIDEA ``check_otp`` would return 0 in a succesfull authentication with the counter 0.
In case of an unsuccessful authentication I think we need to return a negative value.

Reply:
another thing i noticed while reading the header available @ https://github.com/mlankenau/vacman_controller/blob/master/ext/vacman_controller/aal2sdk.h , is the following:

/*********************************************************************/
/* Convert a numeric error code to readable text                     */
/*********************************************************************/
  VDS_EXPORT(aat_ascii *)  AAL2GetErrorMsg(aat_int32 errorNum, aat_ascii *szBuffer);

Should/could that be used the error codes returned? I will try to find more error codes that might be usefull.